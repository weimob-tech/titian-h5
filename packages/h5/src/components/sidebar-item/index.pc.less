@import '../common/static/utils.less';
@import '../common/static/var.less';
@import '../common/static/mixins/index.less';
@sidebar-item:~ '@{project}-sidebar-item';

.@{sidebar-item} {
  display: flex;
  box-sizing: border-box;
  align-items: center;
  padding: var(--sidebar-padding, 34px 20px 34px @gap-28);

  color: var(--sidebar-text-color, @neutral-color-2);
  font-size: var(--sidebar-font-size, @T26);

  &-text {
    z-index: 1;

    line-height: var(--sidebar-line-height, 38px);
    .ellipsis-lines(2);
  }

  // 兼容ios上初次渲染文字缺省不生效
  &-refresh {
    min-width: 0;
  }

  &-active:not(.@{sidebar-item}-disabled) {
    position: relative;

    background: var(--sidebar-active-bg-color, #fff);

    .@{sidebar-item}-text {
      color: var(--sidebar-active-text-color, @brand-color);
      font-weight: @font-weight-bold;
    }

    &::before {
      content: '';

      position: absolute;
      right: 0;
      bottom: ~'calc(-1 * var(--sidebar-radius, (var(--base-radius-size, 0px) + 12px)))';
      width: @sidebar-radius;
      height: @sidebar-radius;

      background: radial-gradient(
        circle at bottom left,
        var(--sidebar-bg-color, @neutral-color-7) 70.7%,
        var(--sidebar-active-bg-color, #fff) 70.7%,
        var(--sidebar-active-bg-color, #fff) 100%
      );
    }

    &::after {
      content: '';

      position: absolute;
      top: ~'calc(-1 * var(--sidebar-radius, (var(--base-radius-size, 0px) + 12px)))';
      right: 0;
      width: @sidebar-radius;
      height: @sidebar-radius;

      background: radial-gradient(
        circle at top left,
        var(--sidebar-bg-color, @neutral-color-7) 70.7%,
        var(--sidebar-active-bg-color, #fff) 70.7%,
        var(--sidebar-active-bg-color, #fff) 100%
      );
    }
  }

  &-disabled {
    color: var(--sidebar-disabled-text-color, @neutral-color-4);
  }
}
