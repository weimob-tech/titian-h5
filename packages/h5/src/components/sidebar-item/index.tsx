import { Component, Method, Element, h, Prop, State } from '@stencil/core';
import { JSXBase } from '@stencil/core/internal';
import { stringToAttrStyle, addShadowRootStyle } from '../common/utils';
import { handle, join } from '../common/utils/namespace';
import { getParent } from '../common/utils/relation';

@Component({
  tag: 'ti-sidebar-item',
  styleUrls: {
    pc: 'index.pc.less',
    h5: 'index.h5.less',
  },
  shadow: true,
})
export class TiSidebarItem {
  @Element() host!: HTMLTiSidebarItemElement;

  @Prop() extCss = '';

  @Prop() label = '';

  @Prop() disabled = false;

  @Prop() badge = '';

  @Prop() dot = false;

  @Prop() extClass?: string = '';

  @Prop() extStyle?: string | JSXBase.HTMLAttributes<Record<string, unknown>>['style'] = '';

  @State() active = false;

  @State() refresh = false;

  get parent() {
    return getParent({
      host: this.host,
      tag: 'ti-sidebar',
      relations: 'ancestor', // 'parent' ,'ancestor',
    })[0];
  }

  componentDidLoad() {
    this.updateDataFromParent();
    this.refresh = true;
  }

  @Method()
  async updateDataFromParent() {
    const { parent } = this;
    if (!parent) {
      return;
    }
    const { children, activeIndex } = await parent.getImperativeHandle();
    const index = children.indexOf(this.host);
    this.active = index === activeIndex;
  }

  onClick = async () => {
    if (this.disabled) return;
    const { parent } = this;
    if (!parent) {
      return;
    }
    const { children } = await parent.getImperativeHandle();
    const index = children.indexOf(this.host);
    parent.setActive(index);
  };

  componentWillLoad() {
    addShadowRootStyle.call(this);
  }

  render() {
    const { active, extClass, extStyle, label, disabled, dot, badge } = this;
    return (
      <div
        class={`${join('sidebar-item', [{ active, disabled }])} ${extClass}`}
        part={extClass}
        style={stringToAttrStyle(extStyle)}
        onClick={this.onClick}
        aria-hidden="true"
      >
        <ti-badge content={badge} dot={dot} atText>
          <div class={handle('sidebar-item', ['text', this.refresh && 'refresh'])}>{label}</div>
        </ti-badge>
      </div>
    );
  }
}
