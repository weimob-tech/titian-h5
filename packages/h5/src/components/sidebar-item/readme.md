# ti-sidebar-item



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type                                   | Default |
| ---------- | ----------- | ----------- | -------------------------------------- | ------- |
| `badge`    | `badge`     |             | `string`                               | `''`    |
| `disabled` | `disabled`  |             | `boolean`                              | `false` |
| `dot`      | `dot`       |             | `boolean`                              | `false` |
| `extClass` | `ext-class` |             | `string`                               | `''`    |
| `extStyle` | `ext-style` |             | `string \| { [key: string]: string; }` | `''`    |
| `label`    | `label`     |             | `string`                               | `''`    |


## Methods

### `updateDataFromParent() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [ti-tree-select](../tree-select)

### Depends on

- [ti-badge](../badge)

### Graph
```mermaid
graph TD;
  ti-sidebar-item --> ti-badge
  ti-badge --> ti-icon
  ti-tree-select --> ti-sidebar-item
  style ti-sidebar-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
