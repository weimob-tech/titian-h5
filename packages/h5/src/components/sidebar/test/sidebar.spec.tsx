import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { TiSidebar } from '../index';
import { TiSidebarItem } from '../../sidebar-item';
describe('渲染 ti-sidebar', () => {
  it('基本用法', async () => {
    const page = await newSpecPage({
      components: [TiSidebar, TiSidebarItem],
      template: () => (
        <ti-sidebar active-index={1}>
          <ti-sidebar-item label="侧边导航" badge="1" />
          <ti-sidebar-item label="IP联名款" />
          <ti-sidebar-item label="精选系列" disabled />
        </ti-sidebar>
      ),
    });
    page.root.setAttribute('active-index', '2');
    const dom = page.root
      .querySelector<HTMLElement>('ti-sidebar-item')
      .shadowRoot.querySelector('.titian-sidebar-item');
    dom?.dispatchEvent(new Event('click'));
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });
});
