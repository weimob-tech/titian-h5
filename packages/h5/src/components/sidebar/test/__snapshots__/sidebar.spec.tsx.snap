// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染 ti-sidebar 基本用法 1`] = `
<ti-sidebar active-index="2">
  <mock:shadow-root>
    <ti-scroll-view class="titian-sidebar" part="" scrolly="">
      <slot></slot>
    </ti-scroll-view>
  </mock:shadow-root>
  <ti-sidebar-item>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-sidebar-item titian-sidebar-item-active" part="">
        <ti-badge attext="" content="1">
          <div class="titian-sidebar-item-refresh titian-sidebar-item-text">
            侧边导航
          </div>
        </ti-badge>
      </div>
    </mock:shadow-root>
  </ti-sidebar-item>
  <ti-sidebar-item>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-sidebar-item" part="">
        <ti-badge attext="" content="">
          <div class="titian-sidebar-item-refresh titian-sidebar-item-text">
            IP联名款
          </div>
        </ti-badge>
      </div>
    </mock:shadow-root>
  </ti-sidebar-item>
  <ti-sidebar-item>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-sidebar-item titian-sidebar-item-disabled" part="">
        <ti-badge attext="" content="">
          <div class="titian-sidebar-item-refresh titian-sidebar-item-text">
            精选系列
          </div>
        </ti-badge>
      </div>
    </mock:shadow-root>
  </ti-sidebar-item>
</ti-sidebar>
`;
